neps/nep-0399.md

```diff
@@ -0,0 +1,375 @@
+---
+NEP: 0399
+Title: Flat Storage
+Author: Aleksandr Logunov <alex.logunov@near.org> Min Zhang <min@near.org>
+DiscussionsTo: https://github.com/nearprotocol/neps/pull/0399
+Status: Draft
+Type: Protocol Track
+Category: Storage
+Created: 30-Sep-2022
+---
+
+## Summary
+
+This NEP proposes the idea of Flat Storage, which stores a flattened map of key/value pairs of the current
+blockchain state on disk. Note that original Trie (persistent merkelized trie) is not removed, but Flat Storage
+allows to make storage reads faster, make storage fees more predictable and potentially decrease them.
+
+## Motivation
+
+Currently, the blockchain state is stored in our storage only in the format of persistent merkelized tries.
+Although it is needed to compute state roots and prove the validity of states, reading from it requires a
+traversal from the trie root to the leaf node that contains the key value pair, which could mean up to 
+2 * key_length disk accesses in the worst case.
+
+In addition, we charge receipts by the number of trie nodes they touched (TTN cost). Note that the number 
+of touched trie node does not always equal to the key length, it depends on the internal trie structure. 
+Based on some feedback from contract developers collected in the past, they are interested in predictable fees,
+but TTN costs are annoying to predict and can lead to unexpected excess of the gas limit. They are also a burden
+for NEAR Protocol client implementations, i.e. nearcore, as exact TTN number must be computed deterministically
+by all clients. This prevents storage optimizations that use other strategies than nearcore uses today.
+
+With Flat Storage, number of disk reads reduces from 2 * key_length to 2, storage read gas fees are simplified 
+by getting rid of TTN cost, and potentially can be reduced because less disk reads are needed.
+
+## Rationale and alternatives
+
+Q: Why is this design the best in the space of possible designs?
+
+A: There are other ideas for how to improve storage performance, such as using 
+  other databases instead of rocksdb, or changing the representation of states
+  to achieve locality of data in the same account. Considering that these ideas 
+  will likely require much more work than FlatStorage, FlatStorage is a good investment
+  of our effort to achieve better storage performances. In addition, the improvement 
+  from FlatStorage can be combined with the improvement brought by these other ideas,
+  so the implementation of FlatStorage won't be rendered obsolete in the future.
+
+Q: What other designs have been considered and what is the rationale for not choosing them?
```

Comment:
one other alternative would be to change the tree type. (for example by moving to something like AVL or B-trees) - 

let's mention pros&cons of this vs flat storage.

Reply:
Done, changed this subsection significantly by listing other ideas: getting rid of state root, increasing TTN.